src/serde_impl.rs

```diff
@@ -93,16 +93,22 @@ macro_rules! impl_serde_for_symbol {
                 &self,
                 serializer: T,
             ) -> ::core::result::Result<T::Ok, T::Error> {
-                self.value.serialize(serializer)
+                // When the value is deserialized below, it will be converted back to a non-zero
+                // type by adding 1 to the value, so we need to account for that here by subtracting
+                // 1 from the value.
+                let zeroable_value = self.value.get() - 1;
+                zeroable_value.serialize(serializer)
             }
         }
 
         impl<'de> ::serde::Deserialize<'de> for $crate::symbol::$name {
             fn deserialize<D: ::serde::Deserializer<'de>>(
                 deserializer: D,
             ) -> ::core::result::Result<Self, D::Error> {
-                let index = <$ty as ::serde::Deserialize<'de>>::deserialize(deserializer)?;
-                let ::core::option::Option::Some(symbol) = Self::new(index) else {
+                let zeroable_value = <$ty as ::serde::Deserialize<'de>>::deserialize(deserializer)?;
+                // Symbol::new() will add 1 to the value to guarantee that the resulting underlying
+                // index will be non-zero.
+                let ::core::option::Option::Some(symbol) = Self::new(zeroable_value) else {
```

Comment:
The previous behavior was correct in my opinion. Non-zero symbol indices _must_ be non-zero. We serialize them to be 0-based (-1) and then we have to add (+1) again indeserialization.

Reply:
Yes, agreed. But just to be clear, there is no behavior change in this diff: I just renamed `index` to `zeroable_value` and added the comment (both of which I now see are incorrect, as per your other comment).